python/triton/compiler/compiler.py

```diff
@@ -453,11 +453,25 @@ def compile(fn, **kwargs):
         name, signature = match.group(1), match.group(2)
         types = re.findall(arg_type_pattern[ir_name], signature)
         if ir_name == 'ttgir':
+            # TODO(jlebar): Using a regex to get num-warps is a hack, and will
+            # break if e.g. someone has an instruction (not module) attribute
+            # named "num-warps".
             num_warps_matches = re.findall(ttgir_num_warps_pattern, src)
             assert len(num_warps_matches) == 1, "Expected exactly one match for num_warps"
-            assert "num_warps" not in kwargs or int(
-                num_warps_matches[0]) == num_warps, "num_warps in ttgir does not match num_warps in compile"
-            num_warps = int(num_warps_matches[0])
+            ir_num_warps = int(num_warps_matches[0])
+            assert "num_warps" not in kwargs or ir_num_warps == num_warps, "num_warps in ttgir does not match num_warps in compile"
+
+            # If warp specialization is enabled, the true number of warps from
+            # the perspective of e.g. CUDA is num-warps times the number of
+            # specialized groups.
+            num_warp_groups_matches = re.findall(r'"triton_gpu.num-warp-groups-per-cta"\s?=\s?(\d+)\s?:', src)
+            assert len(num_warp_groups_matches) == 0 or len(num_warp_groups_matches) == 1, \
+                "Expected triton_gpu.num-warp-groups-per-cta attribute to appear 0 or 1 times"
+            if num_warp_groups_matches:
+                ir_num_warps *= int(num_warp_groups_matches[0])
+
+            num_warps = ir_num_warps
```

Comment:
should we fix it in convert to llvm ir pass instead? It feels a little intrusive to have warp specialization code here.

Reply:
I can do that, but it would be a change in behavior for this code.

Right now we check that the `num_warps` kwarg matches `num-warps` in the module, and *then* we multiply by `num-warp-groups-per-cta`.  If I changed the module's `num-warps` value during convert-to-llvm-ir, then the `num_warps` kwarg would have to change.

I agree this way is a hack, but asking users to pass num_warps == 8 also seems wrong to me, because in the user Python code they actually wrote num_warps = 4.  Like, it's an implementation detail that we're actually launching 8 warps per CUDA block.

WDYT?

Reply:
How was it working before? The WS pass used to change the num_warps in the module?  How come it didn't break things at the time. I think if we can we should fix it in ConvertToLLVM pass in order to keep the workaround more localized but I don't understand how this was working before so maybe it is not possible.

Reply:
Good point.  AFAICT it was broken before as well and people just were not going down this branch.

Reply:
But we do go down this branch now?

Reply:
Apparently not in tests, but "IRL"?  Surely we take this branch unless it's all dead code?

Reply:
my preference would be to leave it out of this PR then. If this was already broken I'm not sure we want to fix it as part of this change. WDYT?

Reply:
Works for me.  I've pushed a change that passes some tests locally, will submit once CI is green.  Thanks again for the review, Thomas.